### DISCUSSION TITLE
spring war방식 + tomcat 관련 collect 문의

### DISCUSSION THREAD

**Discussion by pms131** [naver/scavenger#68] [Q&A] [answered]
안녕하세요 제가 scavenger를 Spring War방식 + tomcat에서 사용하려고 하는데 문의 드립니다.

codebase Scan은 잘 되는데 invoke data가 계속 0으로 잡히고 access시에도 해당 데이터가 수집이 안되는데

혹시 해당 방식을 지원하지 않는건지 아니면 Spring 버전 혹은 경로의 문제인지 문의 드립니다.

- 개발 환경
  - Spring 3.2.13
  - Java 1.8
  - tomcat 8.5.78

**Comment by taeyeon-Kim** [comment]
@pms131 안녕하세요 :)
[configuration](https://github.com/naver/scavenger/blob/develop/doc/installation.md#configuration-2) codebase의 codebase에 webapp path를 지정했을까요??
예를 들면 아래와 같이요
```
codeBase = /tomcat/webapps/ROOT/WEB-INF, ${TOMCAT_HOME}/lib
```

혹시 설정하셨다면 저희가 도와드릴 수 있도록 scavenger configuration 설정 파일과 로그같은 부분도 같이 공유 부탁드립니다.

**Reply by pms131** [reply]
<img width="316" alt="스크린샷 2023-04-26 오후 3 18 22" src="https://user-images.githubusercontent.com/20953536/234486334-194fb1d5-6282-4210-86cd-3551dc91532e.png">

<img width="2784" alt="스크린샷 2023-04-26 오후 3 11 11" src="https://user-images.githubusercontent.com/20953536/234485296-05915055-3abe-4935-8ad3-85b3c8bf353d.png">

로그 : 

26-Apr-2023 15:03:52.067 정보 [main] com.navercorp.scavenger.javaagent.ScavengerAgent.premain [scavenger] scavenger agent version 1.0.4 is starting...
26-Apr-2023 15:03:52.073 정보 [main] com.navercorp.scavenger.javaagent.util.ConfigUtils.locateConfigFile [scavenger] config file found explicitly: /Users/**********/*******/scavenger/infra_web_scavenger.conf


███████╗ ██████╗ █████╗ ██╗   ██╗███████╗███╗   ██╗ ██████╗ ███████╗██████╗
██╔════╝██╔════╝██╔══██╗██║   ██║██╔════╝████╗  ██║██╔════╝ ██╔════╝██╔══██╗
███████╗██║     ███████║██║   ██║█████╗  ██╔██╗ ██║██║  ███╗█████╗  ██████╔╝
╚════██║██║     ██╔══██║╚██╗ ██╔╝██╔══╝  ██║╚██╗██║██║   ██║██╔══╝  ██╔══██╗
███████║╚██████╗██║  ██║ ╚████╔╝ ███████╗██║ ╚████║╚██████╔╝███████╗██║  ██║
╚══════╝ ╚═════╝╚═╝  ╚═╝  ╚═══╝  ╚══════╝╚═╝  ╚═══╝ ╚═════╝ ╚══════╝╚═╝  ╚═╝

                 agent version :: 1.0.4
               config location :: /Users/**********/*******/scavenger/infra_web_scavenger.conf
                       api key :: 5a550a0e-1939-4009-8186-08c3a4c40792
                    server url :: http://localhost:9080
                      app name :: Im_Infra_Web
                   app version :: unspecified
                   environment :: h2
                      codebase :: /Users/**********/*******/im_infra_web/target
                       package :: com.*****.infra.web
             method visibility :: protected
          exclude constructors :: true
      exclude setters, getters :: false
                      hostname :: *****************
     async code base scan mode :: false
     legacy compatibility mode :: false

26-Apr-2023 15:04:10.521 정보 [main] com.navercorp.scavenger.javaagent.collecting.CodeBaseScanner.scan [scavenger] codebase(f4ab931a8c2ad0cb739cff1cb1e4fd519eea9370c94df2ff61338db17d094958) scanned in 18300 ms: 4662 methods
26-Apr-2023 15:04:12.500 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 서버 버전 이름:    Apache Tomcat/8.5.78
26-Apr-2023 15:04:12.500 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log Server 빌드 시각:  Mar 31 2022 16:05:28 UTC
26-Apr-2023 15:04:12.501 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log Server 버전 번호:  8.5.78.0
26-Apr-2023 15:04:12.501 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 운영체제 이름:     Mac OS X
26-Apr-2023 15:04:12.501 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 운영체제 버전:     12.4
26-Apr-2023 15:04:12.501 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 아키텍처:          x86_64
26-Apr-2023 15:04:12.501 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 자바 홈:           /Users/**********/Library/Java/JavaVirtualMachines/corretto-1.8.0_362/Contents/Home/jre
26-Apr-2023 15:04:12.501 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log JVM 버전:          1.8.0_362-b08
26-Apr-2023 15:04:12.501 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log JVM 벤더:          Amazon.com Inc.
26-Apr-2023 15:04:12.501 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log CATALINA_BASE:     /Users/**********/Library/Caches/JetBrains/IntelliJIdea2023.1/tomcat/9cbaca41-3994-4624-af2f-9a3f947fc250
26-Apr-2023 15:04:12.501 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log CATALINA_HOME:     /Users/**********/*******/apache-tomcat-8.5.78
26-Apr-2023 15:04:12.502 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Djava.util.logging.config.file=/Users/**********/Library/Caches/JetBrains/IntelliJIdea2023.1/tomcat/9cbaca41-3994-4624-af2f-9a3f947fc250/conf/logging.properties
26-Apr-2023 15:04:12.503 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager
26-Apr-2023 15:04:12.503 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dvisualvm.id=339610234696875
26-Apr-2023 15:04:12.503 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dvisualgc.id=339610234668083
26-Apr-2023 15:04:12.503 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -agentlib:jdwp=transport=dt_socket,address=127.0.0.1:58040,suspend=y,server=n
26-Apr-2023 15:04:12.503 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -javaagent:/Users/**********/Library/Caches/JetBrains/IntelliJIdea2023.1/captureAgent/debugger-agent.jar
26-Apr-2023 15:04:12.503 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dbil-env=local
26-Apr-2023 15:04:12.503 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dbil-cache=local
26-Apr-2023 15:04:12.503 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dtest-env=local
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -DserverMode=dev
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -javaagent:/Users/**********/*******/scavenger/scavenger-agent-java-1.0.4.jar
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dscavenger.configuration=/Users/**********/*******/scavenger/infra_web_scavenger.conf
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dcom.sun.management.jmxremote=
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dcom.sun.management.jmxremote.port=10993
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dcom.sun.management.jmxremote.ssl=false
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dcom.sun.management.jmxremote.password.file=/Users/**********/Library/Caches/JetBrains/IntelliJIdea2023.1/tomcat/9cbaca41-3994-4624-af2f-9a3f947fc250/jmxremote.password
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dcom.sun.management.jmxremote.access.file=/Users/**********/Library/Caches/JetBrains/IntelliJIdea2023.1/tomcat/9cbaca41-3994-4624-af2f-9a3f947fc250/jmxremote.access
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Djava.rmi.server.hostname=127.0.0.1
26-Apr-2023 15:04:12.504 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Djdk.tls.ephemeralDHKeySize=2048
26-Apr-2023 15:04:12.505 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Djava.protocol.handler.pkgs=org.apache.catalina.webresources
26-Apr-2023 15:04:12.505 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dorg.apache.catalina.security.SecurityListener.UMASK=0027
26-Apr-2023 15:04:12.505 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dignore.endorsed.dirs=
26-Apr-2023 15:04:12.505 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dcatalina.base=/Users/**********/Library/Caches/JetBrains/IntelliJIdea2023.1/tomcat/9cbaca41-3994-4624-af2f-9a3f947fc250
26-Apr-2023 15:04:12.505 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Dcatalina.home=/Users/**********/*******/apache-tomcat-8.5.78
26-Apr-2023 15:04:12.505 정보 [main] org.apache.catalina.startup.VersionLoggerListener.log 명령 행 아규먼트:  -Djava.io.tmpdir=/Users/**********/*******/apache-tomcat-8.5.78/temp


......

26-Apr-2023 15:08:41.468 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.pollDynamicConfig [scavenger] polling dynamic config
26-Apr-2023 15:10:51.467 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.pollDynamicConfig [scavenger] polling dynamic config
26-Apr-2023 15:10:51.606 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.publishInvocationData [scavenger] publishing invocation data: 0 invocations
26-Apr-2023 15:10:51.617 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.publishInvocationData [scavenger] invocation data published
26-Apr-2023 15:13:01.471 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.pollDynamicConfig [scavenger] polling dynamic config
26-Apr-2023 15:13:01.597 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.publishInvocationData [scavenger] publishing invocation data: 0 invocations
26-Apr-2023 15:13:01.616 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.publishInvocationData [scavenger] invocation data published
26-Apr-2023 15:15:11.477 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.pollDynamicConfig [scavenger] polling dynamic config
26-Apr-2023 15:15:12.138 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.publishInvocationData [scavenger] publishing invocation data: 0 invocations
26-Apr-2023 15:15:12.146 정보 [scavenger scheduler] com.navercorp.scavenger.javaagent.publishing.Publisher.publishInvocationData [scavenger] invocation data published



현재 이렇게 설정 되어있습니다.

추가적으로 해당 프로젝트에서 import된 library jar 프로젝트도 혹시 추적 가능한지, 가능하다면 어떤 가이드 참고하면 될지 문의 드립니다.

**Reply by taeyeon-Kim** [reply, marked as answer]
음.. 보기에는 codebase Scan도 잘안되는 것 같은데요.
https://github.com/naver/scavenger/discussions/68#discussioncomment-5727371 에 작성한 것처럼
codeBase는 `APP_PATH/WEB-INF, TOMCAT_PATH/lib` 와 같이 app의 WEB-INF와 tomcat의 lib 절대 경로를 지정해보시겠어요??

> 추가적으로 해당 프로젝트에서 import된 library jar 프로젝트도 혹시 추적 가능한지, 가능하다면 어떤 가이드 참고하면 될지 문의 드립니다.

이 또한 설정한 codebase 범위에 있다면 추적 가능합니다.

**Reply by pms131** [reply]
아 넵 tomcat/ROOT/WEB-INF로 codebase 잡았더니 정상동작 확인했습니다.

그리고 위에도 말씀드렸지만 추가적으로 프로젝트에서 import된 library jar 프로젝트도 혹시 추적 가능한지, 가능하다면 어떤 가이드 참고하면 될지 문의 드려도 될까요??

해당 프로젝트에 다른 java 프로젝트가 jar파일 형태로 Library로 제공되는데, 해당 프로젝트까지 추적하고 싶어서 문의드립니다!
(해당 jar 프로젝트 역시 spring3, java 1.8입니다)

**Reply by taeyeon-Kim** [reply]
넵넵 가능합니다. :) war안에 해당 library가 포함돼있을테니.. 아래처럼 설정해보시겠어요?? 
```
packages=com.*****.infra.web,  com.example.core
```

이 또한 모두 configuration 가이드에 설명 돼 있습니다.
https://github.com/naver/scavenger/blob/develop/doc/installation.md#configuration-2
